### PR TITLE
add additional peakfiletypes in TOPPView

### PR DESCRIPTION
**Pull request by timosachsenberg**
# Description

Additional filetypes in TOPPView.
Addresses #5231 partly

# Checklist:
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.

### CODE REVIEW

File: src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp [hunk 1]
```diff
@@ -134,7 +134,7 @@ namespace OpenMS
   /// supported types which can be opened with File-->Open
   const FileTypes::FileTypeList supported_types({ FileTypes::MZML, FileTypes::MZXML, FileTypes::MZDATA, FileTypes::SQMASS,
                                                   FileTypes::FEATUREXML, FileTypes::CONSENSUSXML, FileTypes::IDXML,
-                                                  FileTypes::DTA, FileTypes::DTA2D,
+                                                  FileTypes::DTA, FileTypes::DTA2D, FileTypes::MGF, FileTypes::MS2,
```
Comment (jpfeuffer) [inline src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:137]: Cool! Can we add MSP for the NIST databases?
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FORMAT/FileHandler.cpp#L369

Or was something speaking against?

Comment (timosachsenberg) [reply src/openms_gui/source/VISUAL/APPLICATIONS/TOPPViewBase.cpp:137]: It can not yet be read via FileHandler. I will take a look.

**Comment by jpfeuffer** [comment]
Are you testing it on your machine? Otherwise, let me know when it is ready to test.

**Comment by timosachsenberg** [comment]
Thread 1 "TOPPView" received signal SIGSEGV, Segmentation fault.
0x00007ffff6583ac0 in OpenMS::StringUtils::toInt(OpenMS::String const&) () from /home/sachsenb/OMS/OpenMS/openms-build/lib/libOpenMS.so
(gdb) bt
#0  0x00007ffff6583ac0 in OpenMS::StringUtils::toInt(OpenMS::String const&) () from /home/sachsenb/OMS/OpenMS/openms-build/lib/libOpenMS.so
#1  0x00007ffff69d7cb5 in OpenMS::MSPFile::load(OpenMS::String const&, std::vector<OpenMS::PeptideIdentification, std::allocator<OpenMS::PeptideIdentification> >&, OpenMS::MSExperiment&) () from /home/sachsenb/OMS/OpenMS/openms-build/lib/libOpenMS.so
#2  0x00007ffff795f6db in OpenMS::TOPPViewBase::addDataFile(OpenMS::String const&, bool, bool, OpenMS::String, unsigned int, unsigned long) ()
   from /home/sachsenb/OMS/OpenMS/openms-build/lib/libOpenMS_GUI.so
#3  0x00007ffff79632bb in OpenMS::TOPPViewBase::openFilesByDialog(OpenMS::String const&) () from /home/sachsenb/OMS/OpenMS/openms-build/lib/libOpenMS_GUI.so
#4  0x00007ffff7addd6c in QtPrivate::QSlotObject<void (OpenMS::TOPPViewBase::*)(OpenMS::String const&), QtPrivate::List<bool>, void>::impl(int, QtPrivate::QSlotObjectBase*, QObject*, void**, bool*) () from /home/sachsenb/OMS/OpenMS/openms-build/lib/libOpenMS_GUI.so

**Comment by timosachsenberg** [comment]
Works and supports files that can be read by MSPGenericFile. Still not ideal that peptide centric MSP files fail to load.

image

**Comment by jpfeuffer** [comment]
Very nice. What is the issue with peptide-centric? Probably some ambiguity in the format, right? Some splitting of strings is empty or you have a string where an int is expected, right?

**Comment by timosachsenberg** [comment]
yes basically a bit different strings and how we store the annotations (pep_id vs in spectrum)

**Comment by timosachsenberg** [comment]
maybe something we could also discuss with @bertrandboudaud

**Comment by bertrandboudaud** [comment]
Hi, i can have a look at that. do you have any example file that fails to load?

**Comment by timosachsenberg** [comment]
There is e.g. https://github.com/OpenMS/OpenMS/blob/develop/src/tests/class_tests/openms/data/MSPFile_test.msp
that I think did not produce reasonable results.
Two things I found we might also look into:
- storing the m/z in the precursor structure of the spectrum
- finding a unified way to store proteomics and metabolomics peak annotations -> if we manage we could add code to visualize them in TOPPView

**Comment by bertrandboudaud** [comment]
In this file MSPFile_test.msp there are 2 issues the MSPGenericFile::load does not handle well

1. "Num peaks" is written instead of "Num Peaks"

2. The way the spectra is organized.
A the moment, MSPGenericFile::load support only 2 kinds of organisation:

```
Num Peaks: 11
35 310; 36 1230; 37 27; 38 303; 47 5240;
66 203; 67 68; 68 77; 82 63; 83 240;
136 350;
```
or:
```
Num Peaks: 11
35:310 36:1230 37:27 38:303 47:5240
66:203 67:68 68:77 82:63 83:240
136:350
```

The test file show this:
```
Num peaks: 14
608.7	974	"y6/0.39 2/2 2.4"
730.4	407	"? 2/2 0.6"
764.4	1788	"b7/0.01 2/2 1.1"
832.0	814	"b8-46/0.53 2/2 1.3"
```

That is 3 fields, tab-separated, and new line for each peak. this organization is not supported at the moment. if I reorganize the file as described above, the file loads sucessfully. we then lose information about the labeling "y6/0.39 2/2 2.4" ...

The documentation about MSP specification I have found is quite old, but it seems that "Num Peaks" with upper case P is what is specified - I imagine we can be less case sensitive - and about the peak data it's not very specific. we can be one line = one peak, tab-separated, or the way it is already supported in MSPGenericFile for the moment which also already valid regarding the specification. However I don't see any information about the possibility of labeling the peak - which is currently not supported at all, it messes the parser as well.

**Comment by jpfeuffer** [comment]
Nice insights. I agree that the specification/standardization is close to non-existent. However, I think it would be really cool if annotations would be supported in any kind of format. I assume when this file was generated, someone had some format in mind at least.

**Comment by bertrandboudaud** [comment]
Looks like this test file was generated from the MSPFile class, which was specialized for proteomics. MSPGenericFile was added after to support metabolomics. It's a bit sad that one cannot read the file created by the other. maybe some factorization can be done there.

**Comment by timosachsenberg** [comment]
I agree. Maybe we could refactor both classes into one MSPFile class.
e.g., you could make (if timer permits) parsing of 
- Num peaks case insensitive
- include something like this into your parser if it makes sense:
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FORMAT/MSPFile.cpp#L159-L169

And we could then give it a go and try to add the proteomics peak annotation format and storing of modifications.

Should already get us close to something that is usable for both proteomics and metabolomics.

**Comment by jpfeuffer** [comment]
I agree. But to be clear, how was this annotation format designed? Is it documented somewhere? Is it use somewhere else than in OpenMS?

**Comment by timosachsenberg** [comment]
I think some parts might be used in the spectrallibrarysearch.
So one way we could make this more universal is to fill MSSpectrum StringDataArray with the peak annotations.
Would work for both peptide and metabolite spectra.
But if we go for this we should discuss what the best way to store them would be.

**Pull request merged**